### PR TITLE
#281 store wasm log

### PR DESCRIPTION
**Pull request by hunshenshi**

### CODE REVIEW

**Comment by Liuhaai** [comment]
`RuntimeLog` model created for logs to be persisted looks strange for me

File: pkg/models/runtime_log.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+// RuntimeLog database model event
+// @def primary                           ID
+// @def unique_index UI_runtime_log_id    RuntimeLogID
+//
+//go:generate toolkit gen model RuntimeLog --database DB
+type RuntimeLog struct {
+	datatypes.PrimaryID
+	RelRuntimeLog
```
Comment (saitofun) [inline pkg/models/runtime_log.go:15]: This field may not be needed

File: pkg/models/runtime_log.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/machinefi/w3bstream/pkg/depends/base/types"
+	"github.com/machinefi/w3bstream/pkg/depends/kit/sqlx/datatypes"
+)
+
+// RuntimeLog database model event
+// @def primary                           ID
+// @def unique_index UI_runtime_log_id    RuntimeLogID
+//
+//go:generate toolkit gen model RuntimeLog --database DB
+type RuntimeLog struct {
```
Comment (saitofun) [inline pkg/models/runtime_log.go:13]: use WasmRuntimeLog?

File: pkg/modules/vm/wasmtime/instance_exports.go [hunk 3]
```diff
@@ -79,22 +73,25 @@ func (ef *ExportFuncs) Log(logLevel, ptr, size int32) int32 {
 	buf, err := ef.rt.Read(ptr, size)
 	if err != nil {
 		ef.log.Error(err)
+		ef.pl.PersLog(conflog.ErrorLevel.String(), err.Error())
 		return wasm.ResultStatusCode_Failed
 	}
-	switch uint32(logLevel) {
-	case logTraceLevel:
+	switch conflog.Level(logLevel) {
```
Comment (saitofun) [inline pkg/modules/vm/wasmtime/instance_exports.go:79]: store all level log to database?

File: pkg/models/runtime_log.go (hunk 1 above)
Comment (hunshenshi) [reply pkg/models/runtime_log.go:15]: `RuntimeLogID` is a unique index, `ProjectName` `AppletName` `SourceName` `InstanceID` may be same

File: pkg/models/runtime_log.go (hunk 2 above)
Comment (hunshenshi) [reply pkg/models/runtime_log.go:13]: RuntimeLog not just wasm log, it also contains host log if need. So I named RuntimeLog.
The feild `SourceName` marked the source of log.

Comment (hunshenshi) [reply pkg/models/runtime_log.go:13]: RuntimeLog not just wasm log, it also contains host log if need. So I named RuntimeLog.
The feild `SourceName` marked the source of log.

File: pkg/modules/vm/wasmtime/instance_exports.go [hunk 4]
```diff
@@ -79,22 +74,25 @@ func (ef *ExportFuncs) Log(logLevel, ptr, size int32) int32 {
 	buf, err := ef.rt.Read(ptr, size)
 	if err != nil {
 		ef.log.Error(err)
+		job.Dispatch(ef.ctx, plog.NewTask(ef.ctx, conflog.ErrorLevel.String(), err.Error()))
```
Comment (Liuhaai) [inline pkg/modules/vm/wasmtime/instance_exports.go:77]: which type of msg to be dumped?

Comment (hunshenshi) [reply pkg/modules/vm/wasmtime/instance_exports.go:77]: all wasm log

Comment (Liuhaai) [reply pkg/modules/vm/wasmtime/instance_exports.go:77]: Then errors from other exported funcs should be recorded as you design.

Comment (Liuhaai) [reply pkg/modules/vm/wasmtime/instance_exports.go:77]: And `vm` module shouldn't know about the async mechanism of the log module. Whether it's print or stored in db should be configured in log module

Comment (hunshenshi) [reply pkg/modules/vm/wasmtime/instance_exports.go:77]: > Then errors from other exported functions should be captured in the design.

Only when the `log` function is called in the wasm code, the wasm runtime log can be stored in the database.

> And the "vm" module should know nothing about the asynchronous mechanism of the log module. Whether it is printed or stored in the database should be configured in the log module.

All wasm runtime logs must be stored in the database for the studio, so they don't need to be configured.

File: pkg/models/runtime_log.go (hunk 2 above)
Comment (saitofun) [reply pkg/models/runtime_log.go:13]: should logs from other modules be stored to their own tables?

**Review by saitofun — APPROVED** [review]

**Pull request merged**